neqo-udp/src/lib.rs

```diff
@@ -7,20 +7,66 @@
 #![allow(clippy::missing_errors_doc)] // Functions simply delegate to tokio and quinn-udp.
 
 use std::{
+    array,
     io::{self, IoSliceMut},
+    iter,
     net::SocketAddr,
     slice::{self, Chunks},
 };
 
+use log::{log_enabled, Level};
 use neqo_common::{qdebug, qtrace, Datagram, IpTos};
 use quinn_udp::{EcnCodepoint, RecvMeta, Transmit, UdpSocketState};
 
-/// Socket receive buffer size.
+/// Receive buffer size
 ///
-/// Allows reading multiple datagrams in a single [`Socket::recv`] call.
-//
-// TODO: Experiment with different values across platforms.
-pub const RECV_BUF_SIZE: usize = u16::MAX as usize;
+/// Fits a maximum size UDP datagram, or, on platforms with segmentation
+/// offloading, multiple smaller datagrams.
+const RECV_BUF_SIZE: usize = u16::MAX as usize;
+
+/// The number of buffers to pass to the OS on [`Socket::recv`].
+///
+/// Platforms without segmentation offloading, i.e. platforms not able to read
+/// multiple datagrams into a single buffer, can benefit from using multiple
+/// buffers instead.
+///
+/// Platforms with segmentation offloading don't benefit from additionally using
+/// multiple buffers.
+///
+/// - Linux/Android: use segmentation offloading via GRO
+/// - Windows: use segmentation offloading via URO (caveat see <https://github.com/quinn-rs/quinn/issues/2041>)
+/// - Apple: no segmentation offloading available, use multiple buffers
+#[cfg(not(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "tvos",
+    target_os = "visionos"
+)))]
+const NUM_BUFS: usize = 1;
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "tvos",
+    target_os = "visionos"
+))]
+// Value approximated based on neqo-bin "Download" benchmark only.
+const NUM_BUFS: usize = 16;
+
+/// A UDP receive buffer.
+pub struct RecvBuf(Vec<Vec<u8>>);
+
+impl RecvBuf {
+    #[must_use]
+    pub fn new() -> Self {
+        Self(vec![vec![0; RECV_BUF_SIZE]; NUM_BUFS])
+    }
+}
+
+impl Default for RecvBuf {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```

Comment:
This seems unused based on the coverage report. Use it or delete it?

Reply:
We enable the [`new_without_default`](https://rust-lang.github.io/rust-clippy/master/index.html#new_without_default) clippy lint, which would complain as such if I remove the `Default` `impl`.

```
warning: you should consider adding a `Default` implementation for `RecvBuf`
  --> neqo-udp/src/lib.rs:60:5
   |
60 | /     pub fn new() -> Self {
61 | |         Self(vec![vec![0; RECV_BUF_SIZE]; NUM_BUFS])
62 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#new_without_default
   = note: `#[warn(clippy::new_without_default)]` on by default
help: try adding this
   |
58 + impl Default for RecvBuf {
59 +     fn default() -> Self {
60 +         Self::new()
61 +     }
62 + }
```